### PR TITLE
Compile all quantifications using a generalized (with specializations) quant algorithm

### PR DESCRIPTION
**Pull request by milseman**
Adds a general purpose quantification algorithm with specializations for common bounds (`0`, `1`, `unbounded`).

Expands testing.

Overhauls some parts of ME's instruction and operand structure.

Note that we continue to suffer from reluctant restores being reentrant, at least as long as we have non-restorable state such as global register values.

### CODE REVIEW

**Comment by milseman** [comment]
Pushed what should be (modulo bugs) all quantification kinds and amounts.

**Comment by milseman** [comment]
@swift-ci please test linux platform

**Review by rxwei — APPROVED** [review]
LGTM with just some minor nits

File: Sources/_MatchingEngine/Engine/Processor.swift [hunk 1]
```diff
@@ -140,13 +142,13 @@ extension Processor {
...
-    // prefix match, transition to accept.
+      // When reaching the end of the match bounds or when we are only doing a
+      // prefix match, transition to accept.
     case (bounds.upperBound, _), (_, .prefix):
       state = .accept
 
-    // When we are doing a full match but did not reach the end of the match
-    // bounds, backtrack if possible.
+      // When we are doing a full match but did not reach the end of the match
+      // bounds, backtrack if possible.
```
[the remark this answers is not in the document]
Comment (milseman) [reply Sources/_MatchingEngine/Engine/Processor.swift:151]: Argh, Xcode!

**Comment by milseman** [comment]
@swift-ci please test linux

**Pull request merged**